[10.x] Fix "after commit" callbacks not running on nested transactions

tonysm opened the pull request:
### Changed

- Deprecates the `Illuminate\Database\DatabaseTransactionsManager::callbackApplicableTransactions` method (no longer used in the framework)
- Deprecates the `Illuminate\Database\DatabaseTransactionsManager::callbacksShouldIgnore` method (no longer used in the framework. We're now using the new `withAfterCommitCallbacksInTestTransactionAwareMode` method, which this one is forwarding to
- Adds a new `Illuminate\Database\DatabaseTransactionsManager::withAfterCommitCallbacksInTestTransactionAwareMode` method meant to be used by the test database traits when they start a transaction. This should make sure that we execute the callbacks properly at the right level of the transaction (since the tests using the database traits start a wrapping transaction and never actually commit)
- Adds a new param to the `Illuminate\Database\DatabaseTransactionsManager::commit($connection, $level = 1)` method. Now, instead of only having to pass the connection name, callers should also pass the transaction level. It's based on that level (and the fact that we're in test mode or not) that we're going to decide if the callbacks should run or not). All calls within the framework were updated.
- Also, the `Illuminate\Database\DatabaseTransactionsManager::commit($connection, $level = 1)` is now called everytime there's a commit. Before it was only being called on the "last one", but now it receives the transaction level and it should decide if the callbacks should be triggered or not.

---

#### A More In-Depth Explanation

Changes the way "after commit" callbacks are executed to avoid
remembering which transactions to ignore.

Before, we remembered the test transaction, so we
could ignore it when deciding to run the "after commit"
callbacks or not.

We're still handling the "after commit" callbacks like that,
but now, instead of remembering which transactions to ignore,
we're always calling the `DatabaseTransactionManager::commit`
method. The difference is that now we're passing the current
transaction level to it. The method will decide whether to run the
callbacks based on that level and whether or not this
is in test mode.

When in tests, instead of setting the current transaction to be
remembered so it could be ignored, we're now only setting the
DatabaseTransactionManager to test mode. When in test mode, it
will execute the callbacks when the transaction count reaches 1
(remember that the test runs in a transaction, so that's the
"root" level). Otherwise, it runs the callbacks when the transaction
level is on level 0 (like in production).

There's also a change in the `DatabaseTransactionManager::addCallback`
method. It now also checks if it's in test mode. When not in test mode,
it only adds the callback to the execution queue if there's an open
transaction. Otherwise, the callback is executed right away. When in
test mode, the number of transactions has to be greater than one for
it to be added to the callbacks queue.

---

#### Context

With the addition of the `withSavepointIfNeeded` in #48144, we added a new transition level that affects tests when you're using "after commit" callbacks and nesting transactions. This wasn't an issue before because we generally have one transaction wrapping. When the `createOrFirst` was added to other `*OrCreate` methods (`firstOrCreate` and `updateOrCreate`), this issue became more apparent.

fixes #48451

tonysm committed:
Tests observer using afterCommit

tonysm committed:
Adds failing test for savepoint and observers using afterCommit

tonysm committed:
Ignore the createOrFirst savepoint to run callbacks

tonysm committed:
Fix typo

tonysm commented on src/Illuminate/Database/Connection.php:
```diff
@@ -1545,6 +1545,18 @@ public function unsetTransactionManager()
+    public function ignoreLatestTransactionForCallbacks()
+    {
+        $this->transactionsManager?->callbacksShouldIgnore(
+            $this->transactionsManager?->getTransactions()?->last()
```
The switch from pulling the `->first()` transaction to `->last()` was intentional. Otherwise, it keeps re-adding the first transition to the list. We want to add the most recent, so we're taking the last instead.

crynobone replied:
In Octane environment, should `$this->transactionsManager` get unset via `disconnect()` method? 

https://github.com/laravel/octane/blob/2.x/src/Listeners/DisconnectFromDatabases.php#L15C26-L15C36

tonysm replied:
Yeah, I've changed the implementation. We won't have to clean it up (it's gone now)

tonysm committed:
Remove DatabaseEloquentAppTest to see if CI passes

tonysm commented:
I've removed the test I added [here](https://github.com/laravel/framework/pull/48466/commits/bf2764bb4acf04d8c1e34a9055d8d78a17f897a3) because the build was failing (even though the test passes locally in isolation - the suite also fails locally when running it entirely). I added that test because I wanted a way to test the framework as an application test was using it. Since this issue only appears in tests, I'm unsure if this is the right way. Ignoring one transaction level was a good solution because only the test code was added to the transition to the ignore list. And now we're calling that inside the framework itself... 🤔

tonysm commented:
I marked it as ready for review so I could get some feedback:
1. Is it okay that the `withSavepointIfNeeded` now adds the savepoint transition to the list of transactions that should be ignored when deciding whether to run the callbacks (using `$afterCommit` on Observers, for instance)
2. Does anyone have some ideas on how to test this? I  added a test [here](https://github.com/laravel/framework/pull/48466/commits/bf2764bb4acf04d8c1e34a9055d8d78a17f897a3), but that broke the build. I see some tests setting the `db.transactions` manually to a mocked like the [EventListenerTest](https://github.com/laravel/framework/blob/10.x/tests/Integration/Events/ListenerTest.php#L20-L50), which might be the way to go. I ran out of time tonight.

crynobone committed:
wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone committed:
wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone commented:
@tonysm added tests based on your example as Integration tests

crynobone committed:
wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone commented on src/Illuminate/Database/DatabaseTransactionsManager.php:
```diff
@@ -14,7 +14,7 @@ class DatabaseTransactionsManager
      * The database transaction that should be ignored by callbacks.
      *
-     * @var \Illuminate\Database\DatabaseTransactionRecord
+     * @var \Illuminate\Support\Collection
```
```suggestion
     * @var \Illuminate\Support\Collection<int, \Illuminate\Database\DatabaseTransactionRecord>
```

mpyw commented:
After careful reconsideration, I believe this corrective approach is incorrect. I consider the following logic to be highly suspect.:

```php
if ($this->transactions == 1) {
    $this->fireConnectionEvent('committing');
    $this->getPdo()->commit();
}

$this->transactions = max(0, $this->transactions - 1);

if ($this->afterCommitCallbacksShouldBeExecuted()) {
    $this->transactionsManager?->commit($this->getName());
}
```

The root cause of the inconsistency is that we are only decrementing `$this->transactions` and not releasing the transaction object from the `TransactionManager`. We need to **merge the registered callbacks into the callback list of the higher-level transaction** and **release the transaction object** at that time. Otherwise, the counts will not match.

tonysm commented:
@crynobone thank you!

taylorotwell committed:
Update src/Illuminate/Database/DatabaseTransactionsManager.php

Co-authored-by: Mior Muhammad Zaki <crynobone@gmail.com>

taylorotwell commented:
I'm starting to worry we are digging a deeper and deeper complicated hole for ourselves after the first `createOrFirst` PR. It seems we keep needing to come back and add additional complicated PRs to fix new bugs we are creating each time.

tonysm commented:
@taylorotwell I feel you. I agree with @mpyw. I think `createOrFirst` on its own was okay. Adding it to `firstOrCreate` made sense to me, but with the complications of the savepoints, I started to think it was perhaps a deeper undertaking than I anticipated. Then, it was added to the `updateOrCreate` method, which seems to bring another set of issues.

But in this case, the issue seems to be with test transactions and the `afterCommit` feature. The savepoints added for `createOrFirst` only surfaced the issue.

This is not an issue in production apps. It only affects testing things that use `afterCommit` callbacks and when using transactions (which is why that feature exists). When `createOrFirst` was added to other methods, it revealed the issue.

I've turned the PR back to draft. I have an idea of how this can be fixed: we could have a test-specific transaction manager that would not count the test wrapping transaction when deciding if the callbacks should run.

tonysm committed:
Changes the way the after commit callbacks are executed
to avoid remembering which transactions to ignore

Before, we were remembering the test transaction so we
could ignore it when deciding to run the after commit
callbacks or not.

We're still handling the after commit callbacks like that
but now instead of remembering which transactions to ignore,
we're always calling the DatabaseTransactionManager::commit
method. The difference is that now we're passing the current
transaction level to it. The method will decide to call the
callbacks or not based on that level and whether or not this
is on in test mode.

When in tests, instead of setting the current transaction to be
remembered so it could be ignored, we're now only setting the
DatabaseTransactionManager to test mode. When in test mode, it
will execute the callbacks when the transactions count reaches
1 (remember that the test runs in a transaction, so that's the
"root" level). Otherwise, it runs the callbacks when the transactions
level is on level 0 (like in production).

There's also a change in the DatabaseTransactionManager::addCallback
method. It now also checks if it's in test mode. When not in test mode,
it only adds the callback to the execution queue if there's an open
transaction. Otherwise, the callback is executed right away. When in
test mode, the number of transactions has to be greater than one for
it to be added to the callbacks queue.

mpyw commented:
@tonysm 

> we could have a test-specific transaction manager

I agree with that!

tonysm committed:
Fix DatabaseTransactionsTest

tonysm committed:
Fix DatabaseTransactionsManagerTest

tonysm committed:
CSFixer

mpyw reviewed (commented):
I'd like to consider the property/method name, but the implementation itself looks good. Greatly appreciate your response!

tonysm committed:
wip

tonysm committed:
Rename method and property and inline usage

tonysm committed:
Adds a depply nested transaction test

crynobone commented on src/Illuminate/Database/DatabaseTransactionsManager.php:
```diff
@@ -54,30 +67,28 @@ public function rollback($connection, $level)
      * @return void
      */
-    public function commit($connection)
+    public function commit($connection, $level)
```
Would it be possible if we set `$level = 1` here since this would be the normal default value except during tests?

tonysm replied:
Hmm, I don't see a value in adding the default value here. The method is only called in from within the `ManagesTransactions` trait inside the framework. Are you worried about other libs calling this method manually? Adding the default value might give the wrong impression that this is optional. If we add a default value of `1` and people call this method without passing the transaction level, we'll always execute the callbacks no matter which transaction level we're currently in... so I'd be careful there. 🤔

crynobone replied:
`db.transaction` binding has been available since Laravel 9, so someone might already be extended or overridden to cater to other use cases such as databases without transactions such as Mongodb, etc. 

Also while it's only called from the class the visibility of the method is still `public`, wouldn't mind much if it was `protected` or `private`.

tonysm replied:
Added

tonysm committed:
Simplify deeply nesting test

tonysm commented:
Ready for review. \cc @taylorotwell @crynobone @mpyw

tonysm committed:
Sets default level value to one (since it's a new parameter)

mpyw commented on src/Illuminate/Database/DatabaseTransactionsManager.php:
```diff
@@ -26,6 +26,19 @@ class DatabaseTransactionsManager
+     *
+     * @return self
+     */
+    public function withAfterCommitCallbacksInTestMode()
```
*`TestMode`* naming seems a bit imprecise. Not all tests use `RefreshDatabase` (and `DatabaseTransactions`), so wouldn't something like ***`DatabaseTransactionsTestMode`*** be more descriptive? Since this method is automatically called by the trait initialization process rather than by the user, I don't think we need to worry about the verbosity of the name.

The other perspective is LGTM!

tonysm replied:
What name do you suggest?

tonysm replied:
Oh, on a re-read, it seems you're suggesting renaming the class. If so, I'm not sure about that. I thought about renaming it to `DatabaseTransactionCallbacksManager` or something, but renaming the class would be a bigger change to the FW itself.

mpyw replied:
No no, I'm just suggesting the method renaming. How about `withAfterCommitCallbacksInDatabaseTransactionsTestMode`? It's a bit verbose but very clear.

tonysm replied:
Done

taylorotwell commented:
@mpyw what are your thoughts on the overall state of this PR now?

mpyw commented:
@taylorotwell 

While I believe it's okay to merge in terms of functionality, there might be room for renaming and structural refactoring. I'd like to leave the final decision up to you.

My thoughts:

### Rename `withAfterCommitCallbacksInTestMode` to `withAfterCommitCallbacksInDatabaseTransactionsTestMode`

Just calling it *"test mode"* gives the impression that it applies to all tests. However, it's more accurate to specify that it's for the *"test mode using the DatabaseTransactions trait."* Even if it's a bit verbose, I thought a more precise naming would be better.

### Separate the test implementation into a different class

Ideally, the implementation used in production and the one for testing should be distinct. Instead of introducing methods like `withAfterCommitCallbacksInDatabaseTransactionsTestMode` to the `Connection` class, I thought it might be better to provide a `TestDatabaseTransactionsManager` (or `DatabaseTransactionsTraitAwareDatabaseTransactionsManager`) for testing purposes and switch the implementation using setter injection. Given that we have a setter injection mechanism in place, I believe it would be more rational to utilize it. The implementation for testing can be easily created by inheriting the production implementation and overriding some of the methods.

mpyw commented:
@taylorotwell 

> ### Separate the test implementation into a different class

Like this:

## `$level == 1 || ($this->afterCommitCallbacksInTestMode && $level == 2))`

### For Production

```php
protected function isLowestLevel(int $level): bool
{
    return $level == 1; // or $level <= 1
}
```

### For `DatabaseTransactions` Trait

```php
protected function isLowestLevel(int $level): bool
{
    return $level == 2; // or $level <= 2
}
```

## `$this->transactions->isEmpty() || ($this->afterCommitCallbacksInTestMode && $this->transactions->count() == 1)`

### For Production

```php
protected function transactionsAreEmpty(): bool
{
    return $this->transactions->isEmpty();
}
```

### For `DatabaseTransactions` Trait

```php
protected function transactionsAreEmpty(): bool
{
    return $this->transactions->count() == 1;
}
```

mpyw commented:
After giving it some thought, I've come to believe that just renaming the method might be sufficient, considering the feasibility. I think the method name `withAfterCommitCallbacksInTestTransactionAwareMode` (and similarly named properties) would be good! Separating the test implementation into a different class doesn't seem realistic, so this should suffice.

tonysm commented:
I'll rename the method in a bit.

tonysm committed:
Rename method

tonysm commented on src/Illuminate/Database/DatabaseTransactionsManager.php:
```diff
@@ -89,36 +100,14 @@ public function commit($connection)
-     * @param  \Illuminate\Database\DatabaseTransactionRecord  $transaction
-     * @return $this
-     */
-    public function callbacksShouldIgnore(DatabaseTransactionRecord $transaction)
```
Should I add back this method signature?

tonysm replied:
I've added this method back and we're marking it as deprecated. It forwards to the new `withAfterCommitCallbacksInTestTransactionAwareMode` method for now.

tonysm commented on src/Illuminate/Database/DatabaseTransactionsManager.php:
```diff
@@ -89,36 +100,14 @@ public function commit($connection)
-     *
-     * @return \Illuminate\Support\Collection
-     */
-    public function callbackApplicableTransactions()
```
This one too, should I add back the signature? We could probably implement it to return all transactions when not in test mode and, if in test mode, skip the first one.

tonysm replied:
I've added this method back with an implementation I believe should work the same as the previous one. We're also marking the method as deprecated

tonysm committed:
Adds back the removed methods from the db.transactions and mark them as deprecated

tonysm committed:
StyleCI

tonysm committed:
Inline the if statement using then collection when() method

mpyw reviewed (commented):
LGTM

tonysm commented:
By the way, if we get one more issue related to `createOrFirst`, I think we should revert from using it in the other `*OrCreate` methods. I can send a PR for that if that happens.

mpyw commented:
I understand your concern, but I really hope that there won't be any more serious issues. The bug this time was due to a lack of consideration on the afterCommit side, so I don't think it's fair to attribute it to the modifications in the `*OrCreate` methods.

I feel this change has benefited many applications. It even obviated the need for a third-party library I used to operate with, so I hope the results of this modification will persist.

tonysm commented:
I also see the benefits of using it in the other methods. It's just that it had some unexpected side effects, unfortunately. No blaming. Just wanted to say that we have that as an option if we have to.

crynobone committed:
wip
